src/unused_finder/graph.rs

```diff
@@ -0,0 +1,104 @@
+use std::{
+    collections::{HashMap, HashSet},
+    rc::Rc,
+};
+
+use super::{
+    node_visitor::{ExportKind, ImportedItem},
+    unused_finder_visitor_runner::ImportExportInfo,
+};
+
+#[derive(Debug, Clone, Default)]
+pub struct GraphFile {
+    pub is_used: bool,
+    // pub package_name: String,
+    pub file_path: String,
+    pub unused_exports: HashSet<ExportKind>,
+    pub import_export_info: ImportExportInfo,
+}
+
+impl GraphFile {
+    pub fn mark_item_as_used(&mut self, item: &ImportedItem) -> bool {
+        let item = ExportKind::from(item);
+        self.unused_exports.remove(&item)
+    }
+}
+
+pub type Edge = (String, String, ImportedItem);
+
+#[derive(Default, Debug)]
+pub struct Graph {
+    pub files: HashMap<String, Rc<GraphFile>>,
+    pub edges: HashSet<Edge>,
+}
+
+impl<'a> Graph {
+    pub fn bfs(&mut self, entries: Vec<String>) -> Vec<String> {
+        let edges = self.get_edges(entries);
+        let mut new_used_stuff = false;
+        let new_entries = edges
+            .iter()
+            .filter_map(|(_, imported, item)| {
+                if let Some(file) = self.files.get_mut(imported) {
+                    if let Some(file) = Rc::get_mut(file) {
+                        if !file.is_used {
+                            file.is_used = true;
+                            new_used_stuff = true;
+                        }
+                        new_used_stuff = new_used_stuff || file.mark_item_as_used(item);
+                        if new_used_stuff {
+                            new_used_stuff = false;
+                            return Some(imported.clone());
+                        }
+                    }
+                }
+                None
+            })
+            .collect();
+        self.edges.extend(edges);
```

Comment:
Can you help me understand why we maintain this self.edges set? It's never consumed as far as I can tell, and its an infinitely-growing set of visited edges. Is this leftover from a previous iteration?

Reply:
Leftover from a previous iteration